File: backend-node/src/services/BlockchainService.ts
import { randomUUID } from "crypto";
import {
  createBlockchainLog,
  getBlockchainLogsByTouristId,
} from "./mongoStore.js";
import { BlockchainLogModel } from "../schemas/index.js";

const MOCK_TX_PREFIX = "0xHACK_SAT_";
const SUCCESS_STATUS = "SUCCESS_ISSUED_ON_TESTNET";

export async function issueDigitalID(touristId: string, idHash: string) {
  return createBlockchainLog({
    touristId,
    dataHash: idHash,
    transactionId: `${MOCK_TX_PREFIX}${randomUUID().slice(0, 8)}`,
    status: SUCCESS_STATUS,
  });
}

export async function verifyIDProof(idHash: string) {
  const log = await BlockchainLogModel.findOne({
    dataHash: idHash,
    status: SUCCESS_STATUS,
  }).lean();
  return !!log;
}

export async function getRecentLogs(touristId: string, limit: number) {
  const logs = await getBlockchainLogsByTouristId(touristId);
  if (limit <= 0 || logs.length <= limit) {
    return logs;
  }
  return logs.slice(0, limit);
}

File: backend-spring/src/main/java/com/safarsathi/service/BlockchainService.java
package com.safarsathi.service;

import com.safarsathi.entity.BlockchainLog;
import com.safarsathi.repository.BlockchainLogRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class BlockchainService {

    private static final String MOCK_TX_PREFIX = "0xHACK_SAT_";
    private static final String SUCCESS_STATUS = "SUCCESS_ISSUED_ON_TESTNET";

    private final BlockchainLogRepository blockchainLogRepository;
    private final SequenceService sequenceService;

    /**
     * Issues a mock digital ID on a simulated blockchain.
     */
    public BlockchainLog issueDigitalID(String touristId, String idHash) {
        BlockchainLog log = BlockchainLog.builder()
                .logId((int) sequenceService.getNextId("blockchainLogId"))
                .touristId(touristId)
                .dataHash(idHash)
                .transactionId(MOCK_TX_PREFIX + UUID.randomUUID().toString().substring(0, 8))
                .status(SUCCESS_STATUS)
                .build();
        return blockchainLogRepository.save(log);
    }

    /**
     * Verify that an ID proof exists on the mock blockchain.
     */
    public boolean verifyIDProof(String idHash) {
        return blockchainLogRepository.findByDataHashAndStatus(idHash, SUCCESS_STATUS) != null;
    }

    /**
     * Get recent blockchain logs for a tourist.
     */
    public List<BlockchainLog> getRecentLogs(String touristId, int limit) {
        List<BlockchainLog> logs = blockchainLogRepository.findByTouristId(
                touristId, Sort.by(Sort.Direction.DESC, "createdAt"));
        if (limit <= 0 || logs.size() <= limit) return logs;
        return logs.subList(0, limit);
    }
}
